File: src/services/aiService.js
const Groq = require("groq-sdk");

const client = new Groq({
  apiKey: process.env.GROQ_API_KEY,
});

const FALLBACK_METADATA = {
  category: "Personal",
  priority: "Medium",
  subtasks: ["Plan task", "Execute task", "Review progress"],
};

/**
 * Generate AI metadata for a task
 * Returns:
 * {
 *   category,
 *   priority,
 *   subtasks
 * }
 */
async function generateTaskMetadata(taskDescription) {
  try {
    const response = await client.chat.completions.create({
      model: "llama-3.3-70b-versatile",

      response_format: { type: "json_object" },

      messages: [
        {
          role: "system",
          content: `
                  You are a task management assistant.
                  Analyze the given task and return STRICT JSON only.

                  Format:
                  {
                    "category": "",
                    "priority": "",
                    "subtasks": ["", "", ""]
                  }

                  Rules:
                  - category must be one of: Work, Personal, Health, Study, Finance, Shopping, Social, Other
                  - priority must be one of: High, Medium, Low
                  - subtasks must be an array of exactly 3 short actionable steps
                  - Return JSON only, no markdown, no explanation
            `,
        },
        {
          role: "user",
          content: `Task: ${taskDescription}`,
        },
      ],

      temperature: 0.3,
    });

    const parsed = JSON.parse(response.choices[0].message.content);

    return {
      category: parsed.category || FALLBACK_METADATA.category,
      priority: parsed.priority || FALLBACK_METADATA.priority,
      subtasks:
        Array.isArray(parsed.subtasks) && parsed.subtasks.length > 0
          ? parsed.subtasks
          : FALLBACK_METADATA.subtasks,
    };
  } catch (e) {
    console.error("generateTaskMetadata Error:", e.message);
    return FALLBACK_METADATA;
  }
}

/**
 * Natural language task parser
 *
 * Example:
 * "Finish Oracle prep next Friday at 6 PM"
 *
 * Returns:
 * {
 *   title,
 *   dueDate,
 *   priority
 * }
 */
async function parseNaturalLanguageTask(inputText) {
  try {
    const response = await client.chat.completions.create({
      model: "llama3-8b-8192",

      response_format: { type: "json_object" },

      messages: [
        {
          role: "system",
          content: `
                  Extract structured task data from the user's input.
                  Return STRICT JSON only, no markdown, no explanation.

                  Format:
                  {
                    "title": "",
                    "dueDate": "",
                    "priority": ""
                  }

                  Rules:
                  - title should be a clean short task title extracted from the input
                  - dueDate should be in ISO 8601 format (e.g. 2026-05-27T18:00:00.000Z), or null if not mentioned
                  - priority must be High, Medium, or Low based on urgency words in the input
                  - Current date for reference: ${new Date().toISOString()}
                  `,
        },
        {
          role: "user",
          content: inputText,
        },
      ],

      temperature: 0.2,
    });

    return JSON.parse(response.choices[0].message.content);
  } catch (error) {
    console.error("parseNaturalLanguageTask Error:", error.message);
    return {
      title: inputText,
      dueDate: null,
      priority: "Medium",
    };
  }
}

/**
 * AI productivity summary
 */
async function generateProductivitySummary(tasks) {
  try {
    const formattedTasks = tasks.map((task) => ({
      title: task.title,
      completed: task.completed,
      priority: task.priority,
      category: task.category,
    }));

    const response = await client.chat.completions.create({
      model: "llama3-8b-8192",

      messages: [
        {
          role: "system",
          content: `
                  You are a productivity coach.
                  Analyze the user's task list and provide:
                  - A short productivity summary
                  - Pending work analysis
                  - One actionable improvement suggestion

                  Keep the response under 120 words. Be concise and encouraging.
                  `,
        },
        {
          role: "user",
          content: JSON.stringify(formattedTasks),
        },
      ],

      temperature: 0.5,
    });

    return response.choices[0].message.content;
  } catch (error) {
    console.error("generateProductivitySummary Error:", error.message);
    return "Unable to generate productivity summary.";
  }
}

module.exports = {
  generateTaskMetadata,
  parseNaturalLanguageTask,
  generateProductivitySummary,
};

File: backend-java/src/main/java/com/gaurav/taskmanager/service/AiService.java
package com.gaurav.taskmanager.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Instant;
import java.util.List;
import java.util.Map;

@Service
public class AiService {

    private static final Logger log = LoggerFactory.getLogger(AiService.class);
    private static final String GROQ_URL = "https://api.groq.com/openai/v1/chat/completions";
    private static final String MODEL = "llama-3.3-70b-versatile";

    @Value("${app.groq.apikey:}") private String apiKey;

    private final HttpClient http = HttpClient.newHttpClient();
    private final ObjectMapper mapper = new ObjectMapper();

    public record AiTaskMetadata(String category, String priority, List<String> subtasks, String title, Instant dueDate) {}

    /**
     * Parse natural language task — returns category, priority, subtasks, title, dueDate.
     * Falls back gracefully when Groq key is absent.
     */
    public AiTaskMetadata parseNaturalLanguageTask(String input) {
        if (apiKey == null || apiKey.isBlank()) {
            log.warn("GROQ_API_KEY not set — returning fallback AI metadata");
            return fallback(input);
        }
        try {
            String systemPrompt = """
                    Extract structured task data from the user's input.
                    Return STRICT JSON only, no markdown, no explanation.

                    Format:
                    {
                      "title": "",
                      "dueDate": "",
                      "priority": "",
                      "category": "",
                      "subtasks": ["", "", ""]
                    }

                    Rules:
                    - title: clean short task title extracted from the input
                    - dueDate: ISO 8601 format or null if not mentioned
                    - priority: High, Medium, or Low based on urgency
                    - category: one of Work, Personal, Health, Study, Finance, Shopping, Social, Other
                    - subtasks: array of exactly 3 short actionable steps
                    - Current date for reference: %s
                    """.formatted(Instant.now());

            String body = mapper.writeValueAsString(Map.of(
                    "model", MODEL,
                    "temperature", 0.2,
                    "response_format", Map.of("type", "json_object"),
                    "messages", List.of(
                            Map.of("role", "system", "content", systemPrompt),
                            Map.of("role", "user", "content", input)
                    )
            ));

            HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(GROQ_URL))
                    .header("Authorization", "Bearer " + apiKey)
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build();

            HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
            JsonNode root = mapper.readTree(resp.body());
            String content = root.at("/choices/0/message/content").asText();
            JsonNode parsed = mapper.readTree(content);

            String title    = parsed.path("title").asText(input);
            String priority = parsed.path("priority").asText("Medium");
            String category = parsed.path("category").asText("Personal");
            String dueDateStr = parsed.path("dueDate").asText(null);
            Instant dueDate = (dueDateStr != null && !dueDateStr.isBlank() && !dueDateStr.equals("null"))
                    ? Instant.parse(dueDateStr) : null;

            List<String> subtasks = List.of();
            JsonNode subtasksNode = parsed.path("subtasks");
            if (subtasksNode.isArray()) {
                subtasks = mapper.convertValue(subtasksNode, mapper.getTypeFactory()
                        .constructCollectionType(List.class, String.class));
            }

            return new AiTaskMetadata(category, priority, subtasks, title, dueDate);

        } catch (Exception e) {
            log.error("AiService error: {}", e.getMessage());
            return fallback(input);
        }
    }

    /**
     * Generate a productivity summary for a list of tasks.
     */
    public String generateProductivitySummary(List<Map<String, Object>> tasks) {
        if (apiKey == null || apiKey.isBlank()) {
            return "Unable to generate productivity summary (GROQ_API_KEY not configured).";
        }
        try {
            String systemPrompt = """
                    You are a productivity coach.
                    Analyze the user's task list and provide:
                    - A short productivity summary
                    - Pending work analysis
                    - One actionable improvement suggestion
                    Keep the response under 120 words. Be concise and encouraging.
                    """;

            String body = mapper.writeValueAsString(Map.of(
                    "model", MODEL,
                    "temperature", 0.5,
                    "messages", List.of(
                            Map.of("role", "system", "content", systemPrompt),
                            Map.of("role", "user", "content", mapper.writeValueAsString(tasks))
                    )
            ));

            HttpRequest req = HttpRequest.newBuilder()
                    .uri(URI.create(GROQ_URL))
                    .header("Authorization", "Bearer " + apiKey)
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(body))
                    .build();

            HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
            JsonNode root = mapper.readTree(resp.body());
            return root.at("/choices/0/message/content").asText("Unable to generate summary.");
        } catch (Exception e) {
            log.error("generateProductivitySummary error: {}", e.getMessage());
            return "Unable to generate productivity summary.";
        }
    }

    private AiTaskMetadata fallback(String input) {
        return new AiTaskMetadata(
                "Personal", "Medium",
                List.of("Plan task", "Execute task", "Review progress"),
                input, null
        );
    }
}
